Adding option 'Host' to BinaryFormat enum + new exampel

Toni-Graphics opened the pull request:
Hi,
I added the enum-varient `Host` to the `BinaryFormat` enum, which automaticly maps under Windows to Coff on MacOS to Macho and on other platforms to Elf.
I also added the translated faerie example (thanks @philipc for the code while he was very nicely helping me out), because i think the exampels which are currently in place, are hard to understand as someone (like me) comming from `faerie`

Bye

philipc commented on src/common.rs:
```diff
@@ -111,6 +111,27 @@ pub enum BinaryFormat {
+}
+
+impl BinaryFormat {
+    /// Returns the BinaryFormat of the enum
+    /// Used for unwraping the `Host` option
+    #[allow(dead_code)]
+    pub fn get(self) -> BinaryFormat {
```
```suggestion
    /// The target's native binary format for relocatable object files.
    ///
    /// Defaults to `Elf` for unknown platforms.
    pub fn native_object() -> BinaryFormat {
```

Instead of adding the `Host` variant, change this to be a method that directly returns the format, so then the user can use `BinaryFormat::target_object()` instead of `BinaryFormat::Host`.

philipc commented on src/write/mod.rs:
```diff
@@ -94,7 +94,10 @@ pub struct Object<'a> {
+    pub fn new(mut format: BinaryFormat, architecture: Architecture, endian: Endianness) -> Object<'a> {
+        if format == BinaryFormat::Host {
+            format = format.get();
+        }
```
This change can be deleted after `Host` is replaced with `native_object`.

philipc commented on src/common.rs:
```diff
@@ -111,6 +111,27 @@ pub enum BinaryFormat {
+    pub fn get(self) -> BinaryFormat {
+        match self {
+            BinaryFormat::Host => {
+                if cfg!(os = "windows") {
```
```suggestion
                if cfg!(target_os = "windows") {
```

philipc commented on src/common.rs:
```diff
@@ -111,6 +111,27 @@ pub enum BinaryFormat {
+            BinaryFormat::Host => {
+                if cfg!(os = "windows") {
+                    BinaryFormat::Coff
+                } else if cfg!(os = "macos") {
```
```suggestion
                } else if cfg!(target_os = "macos") {
```

philipc reviewed (changes requested):
Thanks for the PR!

I'd like a slightly different API for the binary format, see the comments.

Also, please don't add the example in this PR. It was just a rough conversion of the faerie code. I'll add a better example in a different PR.

Toni-Graphics commented:
Ok i integrated your changes but i renamed `native_object `to `host`.
And I also removed (like you requested ) your translated faerie example.
Should i then also remove then the empty example directory?

Toni-Graphics commented:
I fixed it

philipc commented on src/common.rs:
```diff
@@ -113,6 +113,21 @@ pub enum BinaryFormat {
+        } else if cfg!(target_os = "macos") {
+            BinaryFormat::MachO
+        } else {
+           BinaryFormat::Elf
```
```suggestion
            BinaryFormat::Elf
```
